docs/design_proposals/artifact-dependencies.md

```diff
@@ -0,0 +1,366 @@
+- It only reports a single artifact build failure even though there could have been multiple failures.
+
+This will prove misleading after the introduction of artifact dependencies since we can have out of order artifact definitions in the skaffold config which with the current reporting strategy would appear to be building in the wrong order, and also build failures due to failed required artifact builds won't be immediately apparant.
+
+So we introduce a new `BuildLogger` interface as a facade to achieve two things: 
+- Print log messages in the order that it builds.
+> *<ins>Future work</ins>: This however doesn't solve the problem for concurrently running builds as the current skaffold UX can't show parallel statuses. This will need to be addressed separately when we have a different UX with status bars that can show multiple statuses.
+Until then, we can limit the max concurrency to 1 (this is what we currently do anyways)*
+- Report about *all* build failures.
+
+## Image cache
+
+[hash.go](https://github.com/GoogleContainerTools/skaffold/blob/10275c66a142719897894308b9e566953712a0fe/pkg/skaffold/build/cache/hash.go) provides the `getHashForArtifact` function that needs to recursively be called for each of its dependencies and all those values aggregated together would be the hashcode for the artifact. This would ensure that for a cache hit all the cascading dependencies are unchanged. 
+
+> Note: Dependencies provided as environment variables and build args are not resolved yet during hash calculation. That doesn't matter since they are already accounted for above.
+
+Since `cache` package provides a `Builder` implementation it should additionally append all cache hits to the `existing` `Artifact` slice (see [Builder interfaces](#builder-interfaces) above).
+
+## Dev-loop integration
+
+### Build
+
+[dev.go](https://github.com/GoogleContainerTools/skaffold/blob/10275c66a142719897894308b9e566953712a0fe/pkg/skaffold/runner/dev.go#L149) sets up the file monitor callback functions to queue the affected artifact to need rebuild or resync.
+
+Now we'll have to queue the affected artifact along with all the monitored artifacts that are dependent on it and cascade. To do this we'll need the transpose graph of the `Artifact` slice directed graph that we currently have. One way to implement that would be as follows, which is also safe for concurrent access.
+
+```go
+type artifactDAG struct {
+	m *sync.Map
+}
+
+func getArtifactDAG(artifacts []*latest.Artifact) *artifactDAG {
+	dag := &artifactDAG{m: new(sync.Map)}
+	for _, a := range artifacts {
+		for _, d := range a.Dependencies {
+			slice, ok := dag.m.Load(d.ImageName)
+			if !ok {
+				slice = make([]*latest.Artifact, 0)
+			} else {
+				slice = slice.([]*latest.Artifact)
+			}
+			dag.m.Store(d.ImageName, append(slice.([]*latest.Artifact), a))
+		}
+	}
+	return dag
+}
+
+func (dag *artifactDAG) allDependents(artifact *latest.Artifact) []*latest.Artifact {
```

Comment:
Maybe rename to `dependents` since it's not the transitive closure.

Reply:
done